rotkehlchen/exchanges/ftx.py

```diff
@@ -0,0 +1,528 @@
+        try:
+            self._api_query('wallet/all_balances')
+        except RemoteError as e:
+            error = str(e)
+            if 'Not logged in' in error:
+                return False, 'Bad combination of API Keys'
+            raise
+        return True, ''
+
+    def _make_request(
+        self,
+        endpoint: str,
+        start_time: Timestamp = None,
+        end_time: Timestamp = None,
+        limit: int = 100,
+    ) -> Union[List[Dict[str, Any]], Dict[str, List[Any]]]:
+        """Performs an FTX API Query for endpoint
+
+        You can optionally provide extra arguments to the endpoint via the options argument.
+        If this is an ongoing paginating call then provide pagination_next_uri.
+        If you want just the first results then set ignore_pagination to True.
+        """
+        request_verb = "GET"
+        backoff = self.initial_backoff
+
+        # Use a while loop to retry request if rate limit is reached
+        while True:
+            request_url = '/api/' + endpoint
+            options = {}
+            if limit is not None:
+                options['limit'] = limit
+            if start_time is not None:
+                options['start_time'] = start_time
+            if end_time is not None:
+                options['end_time'] = end_time
+
+            if len(options) != 0:
+                request_url += '?' + urlencode(options)
+
+            timestamp = int(time.time() * 1000)
+            signature_payload = f'{timestamp}{request_verb}{request_url}'.encode()
+
+            signature = hmac.new(self.secret, signature_payload, 'sha256').hexdigest()
+            log.debug('FTX API query', request_url=request_url)
+
+            self.session.headers.update({
+                'FTX-KEY': self.api_key,
+                'FTX-SIGN': signature,
+                'FTX-TS': str(timestamp),
+            })
+
+            full_url = self.base_uri + request_url
+            try:
+                response = self.session.get(full_url)
+            except requests.exceptions.RequestException as e:
+                raise RemoteError(f'FTX API request {full_url} failed due to {str(e)}') from e
+
+            if response.status_code == HTTPStatus.TOO_MANY_REQUESTS:
+                if backoff < self.backoff_limit:
+                    log.debug(
+                        f'FTX rate limit exceeded on endpoint {request_url}. Backing off',
+                        seconds=backoff,
+                    )
+                    gevent.sleep(backoff)
+                    backoff = backoff * 2
+                    continue
+            # We got a result here
+            break
+
+        if response.status_code == 401:
+            raise FTXLoginError(
+                f'Authentication to FTX failed for {request_url}. Verify the time of the machine'
+                f'and that the keys for the API are correct',
+            )
+
+        if response.status_code != 200:
```

Comment:
Be consistent. Use the `HTTPStatus.OK` constant here.

Reply:
The error message also mentions that the key might be invalid as the API doesn't provide information about why the unauthorized error was raised. You can make a request to the balances endpoint `ftx.com/api/wallet/balances` with an invalid api key, for example adding a extra char. Then the response has status code 401 ant the response is

```{"success":false,"error":"Not logged in"}```

You can also use the client they provided to verify this https://github.com/ftexchange/ftx/blob/master/rest/client.py

Reply:
I guess you wanted to answer at the other comment.

Sure the same thing can happen for every single exchange. But it will never happen. Why? We verify that the key is correct. If the key is not correct the api key verification call will fail.

From that point and on we use the same key, so this should not happen. Only way for this to happen is for the api key to become invalid somehow in between.

As with every other exchange, this is a `RemoteError`. There is no reason to introduce a new exception and have to handle it everywhere.